src/main/java/com/metamx/common/parsers/JSONPathParser.java

```diff
@@ -0,0 +1,206 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class JSONPathParser implements Parser<String, Object>
+{
+  private final Map<String, Pair<FieldType, JsonPath>> fieldPathMap;
+  private final List<FieldSpec> fieldSpecs;
+  private final boolean useFieldDiscovery;
+  private final ObjectMapper mapper;
+
+  public JSONPathParser(List<FieldSpec> fieldSpecs, boolean useFieldDiscovery, ObjectMapper mapper)
+  {
+    this.fieldSpecs = fieldSpecs;
+    this.fieldPathMap = generateFieldPaths(fieldSpecs);
+    this.useFieldDiscovery = useFieldDiscovery;
+    this.mapper = mapper == null ? new ObjectMapper() : mapper;
+  }
+
+  @Override
+  public List<String> getFieldNames()
+  {
+    return null;
+  }
+
+  @Override
+  public void setFieldNames(Iterable<String> fieldNames)
+  {
+  }
+
+  @Override
+  public Map<String, Object> parse(String input)
+  {
+    try {
+      Map<String, Object> map = new LinkedHashMap<>();
+      Map<String, Object> document = mapper.readValue(input, new TypeReference<Map<String, Object>>() {});
+      for (Map.Entry<String, Pair<FieldType, JsonPath>> entry : fieldPathMap.entrySet()) {
+        String fieldName = entry.getKey();
+        Pair<FieldType, JsonPath> pair = entry.getValue();
+        JsonPath path = pair.rhs;
+        Object parsedVal;
+        if (pair.lhs == FieldType.ROOT) {
+          parsedVal = document.get(fieldName);
+        } else {
+          parsedVal = path.read(document);
+        }
+        if (parsedVal == null) {
+          continue;
+        }
+        parsedVal = valueConversionFunction(parsedVal);
+        map.put(fieldName, parsedVal);
+      }
+      if (useFieldDiscovery) {
+        discoverFields(map, document);
+      }
+      return map;
+    }
+    catch (Exception e) {
+      throw new ParseException(e, "Unable to parse row [%s]", input);
+    }
+  }
+
+  private Map<String, Pair<FieldType, JsonPath>> generateFieldPaths(List<FieldSpec> fieldSpecs)
+  {
+    Map<String, Pair<FieldType, JsonPath>> map = new LinkedHashMap<>();
+    for (FieldSpec fieldSpec : fieldSpecs) {
+      String fieldName = fieldSpec.getName();
+      JsonPath path = JsonPath.compile(fieldSpec.getExpr());
+      Pair<FieldType, JsonPath> pair = new Pair<>(fieldSpec.getType(), path);
+      map.put(fieldName, pair);
+    }
+    return map;
+  }
+
+  private void discoverFields(Map<String, Object> map, Map<String, Object> document)
+  {
+    for (String field : document.keySet()) {
+      if (!map.containsKey(field)) {
+        Object val = document.get(field);
+        if (val == null) {
+          continue;
+        }
+        if (val instanceof Map) {
+          continue;
+        }
+        if (val instanceof List) {
+          if (!isFlatList((List) val)) {
+            continue;
+          }
+        }
+        val = valueConversionFunction(val);
+        map.put(field, val);
+      }
+    }
+  }
+
+  private Object valueConversionFunction(Object val)
```

Comment:
@gianm Conversion function is now applied to List entries, as well as Map entries